File: ios/Plugin/BraintreeCapacitorPlugin.swift
import Foundation
import Capacitor
import BraintreeDropIn
import Braintree

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(BraintreeCapacitorPlugin)
public class BraintreeCapacitorPlugin: CAPPlugin {
    private let implementation = BraintreeCapacitor()
    var mAuthorization: String? = nil
    var isThreeDSecureEnabled = false
    var shouldCollectDeviceData = false
    var isSaveCardCheckBoxVisible = false
    var defaultVaultSetting = false
    var isVaultManagerEnabled = false
    var nameStatus = "Disabled" // "Optional", "Required"
    var disableApplePay = true

    @objc func initialize(_ call: CAPPluginCall) {
        let authKey = call.getString("authorizationKey") ?? ""
        let apiClient = BTAPIClient(authorization: authKey)
        
        if (apiClient == nil) {
            call.reject("Invalid client token. Please ensure your server is generating a valid Braintree ClientToken.")
            return
        } else {
            mAuthorization = authKey
        }
        
        isThreeDSecureEnabled = call
                        .getString("isThreeDSecureEnabled", "false") == "true"
        shouldCollectDeviceData = call
                .getString("shouldCollectDeviceData", "false") == "true"
        isSaveCardCheckBoxVisible = call
                .getString("isSaveCardCheckBoxVisible", "false") == "true"
        defaultVaultSetting = call
                .getString("defaultVaultSetting", "false") == "true"
        isVaultManagerEnabled = call
                .getString("isVaultManagerEnabled", "false") == "true"
        nameStatus = call
            .getString("nameStatus", "Disabled")
        
        disableApplePay = call.getBool("disableApplePay", true) == false ? false : true
        
        
        call.resolve()
    }
    
    @objc func launch(_ call: CAPPluginCall) {
        if (mAuthorization == nil) {
            call.reject("Client token was not provided or was invalid, please ensure you are running the 'initialize' method before using this plugin")
            return
        }
        
        let request =  BTDropInRequest();
        
        request.applePayDisabled = true
        request.venmoDisabled = true
        request.paypalDisabled = true
        request.shouldMaskSecurityCode = true
        request.cardholderNameSetting =
                    getCardholderNameStatus(nameStatus)
        request.vaultCard = defaultVaultSetting
        request.allowVaultCardOverride = isSaveCardCheckBoxVisible
        
        let dropIn = BTDropInController(authorization: mAuthorization ?? "", request: request)
            { (controller, result, error) in
                if (error != nil) {
                    call.reject(error?.localizedDescription ?? "")
                } else if (result?.isCanceled == true) {
                    call.resolve([
                        "status": "cancelled",
                        "requestCode": 0,
                        "details": []
                    ]);
                    
                } else if let result = result {
                    // Use the BTDropInResult properties to update your UI
                    // result.paymentMethodType
                    // result.paymentMethod
                    // result.paymentIcon
                    // result.paymentDescription                    
                    call.resolve(self.createNonceJSObject(result.paymentMethod))
                    
                }
                
                controller.dismiss(animated: true);
            }
        
        if (dropIn != nil) {
            DispatchQueue.main.async {
                self.bridge?.viewController!.present(dropIn!, animated: true, completion: nil)
            }
        }
            
    }
    
    @objc func getLastPaymentMethod(_ call: CAPPluginCall) {
        if (mAuthorization == nil) {
            call.reject("Client token was not provided or was invalid, please ensure you are running the 'initialize' method before using this plugin")
            return
        }
        
        BTDropInResult.mostRecentPaymentMethod(forClientToken: mAuthorization!) { result, error in
            if (error != nil) {
                call.reject(error!.localizedDescription)
            } else if let result = result {
                // Use the BTDropInResult properties to update your UI
                // result.paymentMethodType
                // result.paymentMethod
                // result.paymentIcon
                // result.paymentDescription
                call.resolve(self.createNonceJSObject(result.paymentMethod))
            }
        }
    }
    
    func getCardholderNameStatus(_ status: String) -> BTFormFieldSetting {
        switch (status) {
            case "Optional":
                return BTFormFieldSetting.optional
            case "Required":
                return BTFormFieldSetting.required
            case "Disabled":
                return BTFormFieldSetting.disabled
            default:
                return BTFormFieldSetting.disabled
        }
    }
    
    func createNonceJSObject(_ nonce: BTPaymentMethodNonce?) -> [String: String] {
        var jsNonce: [String: String] = [:]
        if let nonce = nonce {
            jsNonce = [
                "type": nonce.type,
                "description": nonce.description,
                "nonce": nonce.nonce
            ];
            if nonce.isKind(of: BTCardNonce.self) {
                let cardNonce = nonce as! BTCardNonce
                jsNonce["type"] = "Card"
                jsNonce["cardType"] = nonce.type
                jsNonce["lastFour"] = cardNonce.lastFour
                jsNonce["is3DSLiabilityShifted"] =
                    cardNonce.threeDSecureInfo.liabilityShifted ? "true" : "false"
                jsNonce["is3DSLiabilityShiftPossible"] =
                    cardNonce.threeDSecureInfo.liabilityShiftPossible ? "true" : "false"
            } else if nonce.isKind(of: BTPayPalAccountNonce.self) {
                let paypalAccountNonce = nonce as! BTPayPalAccountNonce
                jsNonce["firstName"] = paypalAccountNonce.firstName
                jsNonce["lastName"] = paypalAccountNonce.lastName
                jsNonce["email"] = paypalAccountNonce.email
                jsNonce["phone"] = paypalAccountNonce.phone
                jsNonce["payerId"] = paypalAccountNonce.payerID
                jsNonce["clientMetadataId"] = paypalAccountNonce.clientMetadataID
                jsNonce["billingAddress"] = paypalAccountNonce.billingAddress?.extendedAddress
                jsNonce["shippingAddress"] = paypalAccountNonce.shippingAddress?.extendedAddress
            } else if nonce.isKind(of: BTVenmoAccountNonce.self){
                let venmoAccountNonce = nonce as? BTVenmoAccountNonce
                jsNonce["userName"] = venmoAccountNonce?.username
            } else if nonce.isKind(of: BTApplePayCardNonce.self) {
                // TODO: Retrieve what do you need there
            }
        }
        return jsNonce;
    }
}

File: android/src/main/java/com/parlevelsystems/braintreecapacitorplugin/BraintreeCapacitorPlugin.java
package com.parlevelsystems.braintreecapacitorplugin;
import android.content.Intent;

import androidx.activity.result.ActivityResult;
import com.braintreepayments.api.DropInClient;
import com.braintreepayments.api.DropInRequest;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.ActivityCallback;
import com.getcapacitor.annotation.CapacitorPlugin;


@CapacitorPlugin(name = "BraintreeCapacitor", requestCodes={BraintreeCapacitorPlugin.DROP_IN_REQUEST})
public class BraintreeCapacitorPlugin extends Plugin {

    static final int DROP_IN_REQUEST = 1;
    private BraintreeCapacitor implementation = new BraintreeCapacitor();

    private DropInClient dropInClient;

    private String mAuthorization;
    private boolean isThreeDSecureEnabled = false;
    private boolean shouldCollectDeviceData = false;
    private boolean isSaveCardCheckBoxVisible = false;
    private boolean defaultVaultSetting = false;
    private boolean isVaultManagerEnabled = false;
    private String nameStatus = "Disabled"; // "Optional", "Required"
    private DropInRequest dropInRequest;

    @PluginMethod()
    public void initialize(PluginCall call) {
        mAuthorization = call.getString("authorizationKey", "");
        isThreeDSecureEnabled = call
                .getString("isThreeDSecureEnabled", "false").compareTo("true") == 0;
        shouldCollectDeviceData = call
                .getString("shouldCollectDeviceData", "false").compareTo("true") == 0;
        isSaveCardCheckBoxVisible = call
                .getString("isSaveCardCheckBoxVisible", "false").compareTo("true") == 0;
        defaultVaultSetting = call
                .getString("defaultVaultSetting", "false").compareTo("true") == 0;
        isVaultManagerEnabled = call
                .getString("isVaultManagerEnabled", "false").compareTo("true") == 0;
        nameStatus = call
                .getString("nameStatus", "Disabled");

        if (mAuthorization == null || mAuthorization.isEmpty()) {
            call.reject("Client token was not provided or was invalid,"
                    + " please ensure you are running the 'initialize' method"
                    + " before using this plugin");
            return;
        }


        call.resolve();
    }

    @PluginMethod()
    public void launch(PluginCall call) {
        if (mAuthorization == null || mAuthorization.isEmpty()) {
            call.reject("Client token was not provided or was invalid,"
                    + " please ensure you are running the 'initialize' method"
                    + " before using this plugin");
            return;
        }

        Intent intent = new Intent(
            getActivity().getApplicationContext(),
            BraintreeDropinActivity.class
        );

        intent.putExtra("authorization", mAuthorization);
        intent.putExtra("isThreeDSecureEnabled", isThreeDSecureEnabled == true ? "true": "false");
        intent.putExtra("isSaveCardCheckBoxVisible", isSaveCardCheckBoxVisible == true ? "true": "false");
        intent.putExtra("defaultVaultSetting", defaultVaultSetting == true ? "true": "false");
        intent.putExtra("isVaultManagerEnabled", isVaultManagerEnabled == true ? "true": "false");
        intent.putExtra("nameStatus", nameStatus);

        startActivityForResult(call, intent, "checkoutResult");

    }

    @PluginMethod()
    public void getLastPaymentMethod(final PluginCall call) {
        call.reject("No possible to get elements");
        return;
        
    }

    @ActivityCallback
    private void checkoutResult(PluginCall call, ActivityResult result) {
        Intent intent = result.getData();
        String json = intent.getStringExtra("result");
        try {
            JSObject response = new JSObject(json);
            call.resolve(response);
        } catch (Exception e) {
            call.reject(e.getMessage(), e);
        }
    }


}
